WIP fix(tags)!: reduce tag filter query from 3+ second

luceos opened the pull request:
This PR removes the subqueries inserted for each selected tag. As this
caused the query to take over 3 seconds on communities with hundred
million posts.

```sql
and (`discussions`.`id` in (select `discussion_id` from `discussion_tag` where `tag_id` = 105))
```

The above part takes on average 3.85 seconds, with a join it takes 0.00
seconds.

BREAKING CHANGE: extending the query might have different outcome as we
now use a join and no subquery.

**Necessity**

- [ ] Has the problem that is being solved here been clearly explained?
- [ ] If applicable, have various options for solving this problem been considered?
- [ ] For core PRs, does this need to be in core, or could it be in an extension?
- [ ] Are we willing to maintain this for years / potentially forever?

**Confirmed**

- [ ] Frontend changes: tested on a local Flarum installation.
- [ ] Backend changes: tests are green (run `composer test`).
- [ ] Core developer confirmed locally this works as intended.
- [ ] Tests have been added, or are not appropriate here.

**Required changes:**

- [ ] Related documentation PR: (Remove if irrelevant)
- [ ] Related core extension PRs: (Remove if irrelevant)

---



The full query that caused us to investigate:

```sql
select `discussions`.*
from `discussions`
where (`discussions`.`id` not in (select `discussion_id`
                                  from `discussion_tag`
                                  where `tag_id` not in (select `tags`.`id`
                                                         from `tags`
                                                         where (`tags`.`id` in (select `perm_tags`.`id`
                                                                                from `tags` as `perm_tags`
                                                                                where (`perm_tags`.`is_restricted` = 1 and
                                                                                       `perm_tags`.`id` in ('202', '205', '206', '89'))
                                                                                   or `perm_tags`.`is_restricted` = 0) and
                                                                (`tags`.`parent_id` in (select `perm_tags`.`id`
                                                                                        from `tags` as `perm_tags`
                                                                                        where (`perm_tags`.`is_restricted` = 1 and
                                                                                               `perm_tags`.`id` in ('202', '205', '206', '89'))
                                                                                           or `perm_tags`.`is_restricted` = 0) or
                                                                 `tags`.`parent_id` is null)))))
  and (`discussions`.`is_private` = 0 or (((`discussions`.`is_approved` = 0 and (`discussions`.`user_id` = 106739 or
                                                                                 ((`discussions`.`id` not in
                                                                                   (select `discussion_id`
                                                                                    from `discussion_tag`
                                                                                    where `tag_id` not in
                                                                                          (select `tags`.`id`
                                                                                           from `tags`
                                                                                           where (`tags`.`id` in
                                                                                                  (select `perm_tags`.`id`
                                                                                                   from `tags` as `perm_tags`
                                                                                                   where (`perm_tags`.`is_restricted` = 1 and `perm_tags`.`id` in ('89'))) and
                                                                                                  (`tags`.`parent_id` in
                                                                                                   (select `perm_tags`.`id`
                                                                                                    from `tags` as `perm_tags`
                                                                                                    where (`perm_tags`.`is_restricted` = 1 and `perm_tags`.`id` in ('89'))) or
                                                                                                   `tags`.`parent_id` is null))))) and
                                                                                  exists(select *
                                                                                         from `tags`
                                                                                                  inner join `discussion_tag` on `tags`.`id` = `discussion_tag`.`tag_id`
                                                                                         where `discussions`.`id` = `discussion_tag`.`discussion_id`)))) or
                                           (`discussions`.`id` in (select distinct `recipients`.`discussion_id`
                                                                   from `recipients`
                                                                   where `recipients`.`removed_at` is null
                                                                     and (`recipients`.`user_id` = 106739))))))
  and (`discussions`.`hidden_at` is null or `discussions`.`user_id` = 106739 or
       ((`discussions`.`id` not in (select `discussion_id`
                                    from `discussion_tag`
                                    where `tag_id` not in (select `tags`.`id`
                                                           from `tags`
                                                           where (`tags`.`id` in (select `perm_tags`.`id`
                                                                                  from `tags` as `perm_tags`
                                                                                  where (`perm_tags`.`is_restricted` = 1 and 0 = 1)) and
                                                                  (`tags`.`parent_id` in (select `perm_tags`.`id`
                                                                                          from `tags` as `perm_tags`
                                                                                          where (`perm_tags`.`is_restricted` = 1 and 0 = 1)) or
                                                                   `tags`.`parent_id` is null))))) and exists(select *
                                                                                                              from `tags`
                                                                                                                       inner join `discussion_tag` on `tags`.`id` = `discussion_tag`.`tag_id`
                                                                                                              where `discussions`.`id` = `discussion_tag`.`discussion_id`)))
  and (`discussions`.`comment_count` > 0 or `discussions`.`user_id` = 106739 or
       ((`discussions`.`id` not in (select `discussion_id`
                                    from `discussion_tag`
                                    where `tag_id` not in (select `tags`.`id`
                                                           from `tags`
                                                           where (`tags`.`id` in (select `perm_tags`.`id`
                                                                                  from `tags` as `perm_tags`
                                                                                  where (`perm_tags`.`is_restricted` = 1 and 0 = 1)) and
                                                                  (`tags`.`parent_id` in (select `perm_tags`.`id`
                                                                                          from `tags` as `perm_tags`
                                                                                          where (`perm_tags`.`is_restricted` = 1 and 0 = 1)) or
                                                                   `tags`.`parent_id` is null))))) and exists(select *
                                                                                                              from `tags`
                                                                                                                       inner join `discussion_tag` on `tags`.`id` = `discussion_tag`.`tag_id`
                                                                                                              where `discussions`.`id` = `discussion_tag`.`discussion_id`)))
  and (`discussions`.`id` in (select `discussion_id` from `discussion_tag` where `tag_id` = 193))
order by `is_stickiest` desc, `is_sticky` desc, `last_posted_at` desc
limit 21;
```

imorland committed:
Bump JS deps

imorland committed:
Update JS imports

imorland committed:
Format JS

flarum-bot committed:
Bundled output for commit 30b2fcce665332909bf5da6be53e0f823b649219

Includes transpiled JS/TS.

[skip ci]

imorland committed:
[chore] Update JS imports (#80)

flarum-bot committed:
Bundled output for commit 0d9c95b0960c5a260c569de43bdc39682a8c56de

Includes transpiled JS/TS.

[skip ci]

askvortsov1 committed:
fix: Ensure `user` not false in post component

If no information about a relation exists, it will return false, not undefined. Therefore, the nullsafe operator is not enough to avoid issues here.

askvortsov1 committed:
fix: don't spellcheck custom LESS

This should stop browsers from "correcting" code, and from converting regular quotes to "smart" quotes.

askvortsov1 committed:
fix: Allow outside click on modal focus trap

Without this, alerts triggered while a modal is open can't be dismissed.

askvortsov1 committed:
Fix frontend branch name

flarum-bot committed:
Bundled output for commit 03aa76cacef96af682b18ec762f5f849ce550a79

Includes transpiled JS/TS, and Typescript declaration files (typings).

[skip ci]

flarum-bot committed:
Bundled output for commit 268d44e89dd5bfd28f69ceb9f4a76557d17e87ed

Includes transpiled JS/TS, and Typescript declaration files (typings).

[skip ci]

imorland committed:
chore: repo maintenance (#5)

* Bump deps

* Update imports

* Format

imorland committed:
chore: repo maintenance (#43)

flarum-bot committed:
Bundled output for commit d593d3ca159278bfea6721589ead09369d0f2fca

Includes transpiled JS/TS.

[skip ci]

imorland committed:
chore: repo maintenance (#30)

* Bump deps

* Update imports

* Format

flarum-bot committed:
Bundled output for commit f442a43205eacdfea6ce1c1017b6e0634aa1b070

Includes transpiled JS/TS.

[skip ci]

imorland committed:
chore: repo maintenance (#30)

* Bump deps

* Update imports

* Format

flarum-bot committed:
Bundled output for commit 542fac888d3aa5eedc9d437381caf101e8e4c1a5

Includes transpiled JS/TS.

[skip ci]

imorland committed:
chore: repo maintenance (#12)

* Bump deps

* Update imports

* Format

imorland committed:
chore: repo maintenance (#28)

* Bump deps

* Update imports

* format

flarum-bot committed:
Bundled output for commit 7b621df353d1ece1a2880ba74d68eb1ba4240605

Includes transpiled JS/TS.

[skip ci]

flarum-bot committed:
Bundled output for commit 200d48820996207938b3732cd87336d1b8648c2d

Includes transpiled JS/TS.

[skip ci]

imorland committed:
chore: repo maintenance (#32)

* Bump deps

* Update imports

imorland committed:
chore: repo maintenance (#16)

* Bump deps

* Update namespaces

* Format

flarum-bot committed:
Bundled output for commit aba58f453a323cd81d86dd36ce064ab44da0a35a

Includes transpiled JS/TS.

[skip ci]

flarum-bot committed:
Bundled output for commit b07c98b5d983cfdd911386ed46ce313002ed7db7

Includes transpiled JS/TS.

[skip ci]

askvortsov1 committed:
fix: return user when creating flags (#42)

Fixes https://github.com/flarum/QualityAssurance/issues/41

When the API response is ingested by `pushObject`, since the flag instance is new, there is nothing to merge with, and therefore, there's no user on the flag instance in the store. As a result, the flag is displayed with "deleted" as the creating user until a page refresh occurs.

More broadly, this could be fixed by either:
- Refactoring the JSON API layer so that all singular operations return the same set of includes.
- Merging the API Paylod with sent data on model save (for new instances only).
But the former would require a major breaking rewrite, and the latter could introduce potentially unwanted side effects, and is effectively a bandaid over inconsistent backend behavior.

askvortsov1 committed:
Some fixes to tag metadata calculations (#154)

* If a discussion doesn't exist, it can't be the last posted discussion

This fixes some rare errors where deleting the only post in the last posted discussion of a tag triggers a DB integrity exception.

* Decrement tag's discussion count when deleting a post deletes an unhidden discussion.

Co-authored-by: Sami Mazouz <sychocouldy@gmail.com>

davwheat committed:
fix: mark `near` as optional in discussion route (#3257)

flarum-bot committed:
Bundled output for commit 59a5871b59bded4019efdaca1beac7f04c8fa36e

Includes transpiled JS/TS, and Typescript declaration files (typings).

[skip ci]

askvortsov1 committed:
chore: bump @askvortsov/rich-icu-message-formatter (#3264)

Fixes https://github.com/flarum/core/issues/3259

flarum-bot committed:
Bundled output for commit c04f4c05f72962069a1db4fea3857f1aa561e67d

Includes transpiled JS/TS, and Typescript declaration files (typings).

[skip ci]

davwheat committed:
chore: switch to Webpack 5 (#35)

flarum-bot committed:
Bundled output for commit 3d864fd413a0ed3dff1a0b75009fd8267b2fe595

Includes transpiled JS/TS.

[skip ci]

rafaucau committed:
Delete package-lock.json (#26)

askvortsov1 committed:
chore: flarum-cli audit infra --fix

askvortsov1 committed:
chore: flarum-cli audit infra --fix

flarum-bot committed:
Bundled output for commit 05246fb155cd43dba1e71caa5d481c84509c00e4

Includes transpiled JS/TS.

[skip ci]

askvortsov1 committed:
chore: flarum-cli audit infra --fix

askvortsov1 committed:
chore: fl-dev audit infra --fix

askvortsov1 committed:
chore: remove unnecessary tsconfig

askvortsov1 committed:
Merge branch 'master' of github.com:flarum/approval

askvortsov1 committed:
chore: flarum-cli audit infra --fix

flarum-bot committed:
Bundled output for commit 5bb0d6cb43505b5355ec88c20674f0bce36525ef

Includes transpiled JS/TS, and Typescript declaration files (typings).

[skip ci]

flarum-bot committed:
Bundled output for commit 99033734e2d815cc625a3a61d363d5c52857856a

Includes transpiled JS/TS.

[skip ci]

askvortsov1 committed:
chore: flarum-cli audit infra --fix

flarum-bot committed:
Bundled output for commit 3858ebdd8617718a0a6e0416ca64218cbf4e277b

Includes transpiled JS/TS.

[skip ci]

askvortsov1 committed:
chore: flarum-cli audit infra --fix

askvortsov1 committed:
chore: flarum-cli audit infra --fix

askvortsov1 committed:
chore: flarum-cli audit infra --fix

flarum-bot committed:
Bundled output for commit f5c1a0282d65624be9343f290cfffa883afa0cd4

Includes transpiled JS/TS.

[skip ci]

askvortsov1 committed:
chore: flarum-cli audit infra --fix

flarum-bot committed:
Bundled output for commit 2c0e6798f96a0e24f5ae0357e604a6e30ef1e5bc

Includes transpiled JS/TS.

[skip ci]

askvortsov1 committed:
chore: flarum-cli audit infra --fix

askvortsov1 committed:
chore: flarum-cli audit infra --fix

flarum-bot committed:
Bundled output for commit e9feb0aa58005e5800438cf066851f2c05cc95ff

Includes transpiled JS/TS.

[skip ci]

askvortsov1 committed:
chore: flarum-cli audit infra --fix

flarum-bot committed:
Bundled output for commit 5f43267a400effede244aa7ceeae5d643773c0c1

Includes transpiled JS/TS.

[skip ci]

askvortsov1 committed:
chore: flarum-cli audit infra --fix

askvortsov1 committed:
chore: flarum-cli audit infra --fix

askvortsov1 committed:
chore: flarum-cli audit infra --fix

flarum-bot committed:
Bundled output for commit 5c6f17d5086447023a38f848de7f79d80922d3e8

Includes transpiled JS/TS.

[skip ci]

askvortsov1 committed:
chore: flarum-cli audit infra --fix

askvortsov1 committed:
chore: flarum-cli audit infra --fix

flarum-bot committed:
Bundled output for commit de4a0db0eca4541b5aca362546fcce524db4e4d6

Includes transpiled JS/TS.

[skip ci]

flarum-bot committed:
Bundled output for commit e03273e3686928418c8cef8e6b5f8515a4d7bcaf

Includes transpiled JS/TS.

[skip ci]

askvortsov1 committed:
chore: flarum-cli audit infra --fix

flarum-bot committed:
Bundled output for commit b363cd18c2eb6c8039ba861539b5da1db39ed217

Includes transpiled JS/TS.

[skip ci]

askvortsov1 committed:
fix: typings generation script

flarum-bot committed:
Bundled output for commit 9ba882619c50652ca5566442d8d9d44b37400873

Includes transpiled JS/TS, and Typescript declaration files (typings).

[skip ci]

askvortsov1 committed:
fix: typings generation script

askvortsov1 committed:
fix: typings generation script

flarum-bot committed:
Bundled output for commit 85e6076553c504f56e67d948748bedc6660b847f

Includes transpiled JS/TS, and Typescript declaration files (typings).

[skip ci]

askvortsov1 committed:
fix: build typings script

askvortsov1 committed:
fix: build typings script

askvortsov1 committed:
fix: build typings script

askvortsov1 committed:
Merge branch 'master' of github.com:flarum/pusher

askvortsov1 committed:
Merge branch 'master' of github.com:flarum/akismet

askvortsov1 committed:
chore: bump rich icu message formatter

flarum-bot committed:
Bundled output for commit 46a7b3419b50d28b2cc3700b29824262af6622a3

Includes transpiled JS/TS, and Typescript declaration files (typings).

[skip ci]

luceos committed:
Release v1.2

luceos committed:
Release v1.2

luceos committed:
Release v1.2

luceos committed:
Release v1.2

luceos committed:
Release v1.2

luceos committed:
Release v1.2

luceos committed:
Release v1.2

luceos committed:
Release v1.2

luceos committed:
Release v1.2

luceos committed:
Release v1.2

luceos committed:
Release v1.2

luceos committed:
Release v1.2

luceos committed:
Release v1.2

luceos committed:
Release v1.2

luceos committed:
Release v1.2

luceos committed:
Release v1.2

luceos committed:
Release v1.2

luceos committed:
Release v1.2

luceos committed:
Release v1.2

luceos committed:
fixing error in tests for missing timezone

luceos committed:
update changelog

luceos committed:
Update Application.php

update constant for next version

askvortsov1 committed:
chore: bump rich text ICU Message formatter

This fixes https://discuss.flarum.org/d/29914-utf-encoding-error-in-title/6

flarum-bot committed:
Bundled output for commit a0c5bc08280e1bd503a9fa1ac2762137ed31b9ac

Includes transpiled JS/TS, and Typescript declaration files (typings).

[skip ci]

askvortsov1 committed:
chore: flarum-cli audit infra --fix

askvortsov1 committed:
fix: allow path repos

Temporarily configure cwd while running composer commands so that path repositories can be resolved properly.

Fixes https://github.com/flarum/package-manager/issues/6

askvortsov1 committed:
chore: bump yarn lock

askvortsov1 committed:
chore: exclude PHP 8.1 tests

These prompt deprecation warnings that break CI. We need to fix those, but that's a problem for a later time.

askvortsov1 committed:
fix: suppress xdebug warnings

askvortsov1 committed:
Merge pull request #12 from flarum/as/monorepo-prep

Monorepo Prep

askvortsov1 committed:
chore: add scaffolding exclude config

This ensures `flarum-cli audit infra --fix` works.

askvortsov1 committed:
chore: flarum-cli audit infra --fix

askvortsov1 committed:
chore: convert to TypeScript

flarum-bot committed:
Bundled output for commit f3d0bdce47cc27f96e69b9c44197dd0acccd2220

Includes transpiled JS/TS, and Typescript declaration files (typings).

[skip ci]

askvortsov1 committed:
fix: minor TS improvements

askvortsov1 committed:
chore: update yarn lockfile

flarum-bot committed:
Bundled output for commit 6d3da9c5f2f18277f439a751fb96e5a80e7b1205

Includes transpiled JS/TS, and Typescript declaration files (typings).

[skip ci]

askvortsov1 committed:
chore: flarum-cli audit infra --fix

askvortsov1 committed:
chore: drop unnecessary `ci` script

askvortsov1 committed:
chore: regenerate yarn.lock

flarum-bot committed:
Bundled output for commit 07fd988dbd0c116eb98144ffd845644121ae4f07

Includes transpiled JS/TS, and Typescript declaration files (typings).

[skip ci]

rafaucau committed:
feat: use `ColorPreviewInput` in the `EditTagModal` (#157)

flarum-bot committed:
Bundled output for commit eed25d18cb55724520ef172d779617188aa7b6c6

Includes transpiled JS/TS.

[skip ci]

davwheat committed:
chore: remove call to `console.log`

flarum-bot committed:
Bundled output for commit 2c0c5c01a27ab3d2fb319c8951bde46b6fad7820

Includes transpiled JS/TS, and Typescript declaration files (typings).

[skip ci]

davwheat committed:
fix: composer header overlap with controls

davwheat committed:
fix: overlap in minimized composer

luceos committed:
fix: missing import for post filterer

zerosonesfun committed:
fix(tag label): tag text misaligned in iOS Safari (#160)

askvortsov1 committed:
fix: regenerate lockfile

This should fix some JS build issues.

askvortsov1 committed:
fix: regenerate lockfile

This should fix some build issues.

askvortsov1 committed:
fix: regenerate lockfile

This should help fix some build issues.

askvortsov1 committed:
chore: make translator / rich translator typings available to exts

davwheat committed:
fix: sort posts by number instead of creation datetime (#3282)

Co-authored-by: Alexander Skvortsov <38059171+askvortsov1@users.noreply.github.com>

flarum-bot committed:
Bundled output for commit 397e4786da109282048bc3a8be5eb81be02ab0e4

Includes transpiled JS/TS, and Typescript declaration files (typings).

[skip ci]

askvortsov1 committed:
fix: refer to bundled dist-typing types instead of src ones

This ensures that extensions referring to core's built dist typings don't try to pull from core's src code.

flarum-bot committed:
Bundled output for commit 351410677fc8b94d0380bf7be015b939947941d2

Includes transpiled JS/TS, and Typescript declaration files (typings).

[skip ci]

clarkwinkelmann committed:
Make frontend use gate output instead of hard-coded logic
Hide nickname from edit modal if user cannot edit

clarkwinkelmann committed:
Fix permission check in edit modal
Previously was checking ability to edit own user instead of user selected in modal

clarkwinkelmann committed:
Fix active search jumping to last item during loading

davwheat committed:
fix: don't submit form when search clear button clicked (#3260)

flarum-bot committed:
Bundled output for commit 567c0a0d7a35469dc3faf44f1c8a9e130ddc8c6f

Includes transpiled JS/TS, and Typescript declaration files (typings).

[skip ci]

clarkwinkelmann committed:
Fix pinned pane broken after discussion creation
The pane would appear but the page content would not shift to the right because the hasPane className would be missing

flarum-bot committed:
Bundled output for commit dba6ad9d571c644dfae9ea4e7396e7f2ebc0ba80

Includes transpiled JS/TS, and Typescript declaration files (typings).

[skip ci]

flarum-bot committed:
Bundled output for commit 05e936bbe4220e46789a21f40897a0a89fe3da7a

Includes transpiled JS/TS.

[skip ci]

imorland committed:
[chore] update deprecated event comments (#41)

askvortsov1 committed:
feat: add compiled dist-typings folder

imorland committed:
fix: unread filter gambit does not return results when > 999 read discussions for a user (#3295)

therealsujitk committed:
Add toolbar support for admin panel (#36)

flarum-bot committed:
Bundled output for commit 16d5cc11e3aee5c94aeed877987cdb199a2a0d2c

Includes transpiled JS/TS.

[skip ci]

dsevillamartin committed:
Replace ColorPreviewInput preview box with 'color' input (#3271)

flarum-bot committed:
Bundled output for commit 904deda6b9cdfb2d9527c2b22f0e589b4f9bd5e8

Includes transpiled JS/TS, and Typescript declaration files (typings).

[skip ci]

dsevillamartin committed:
Fix tag page mismatches if accents/capitalizations differ in slug (#161)

Fixes flarum/core#3286

flarum-bot committed:
Bundled output for commit 671b3f4e54176edc2b8d36321d75e6d135d114be

Includes transpiled JS/TS.

[skip ci]

askvortsov1 committed:
Merge remote-tracking branch 'js_packages_webpack-config/REWRITE'

askvortsov1 committed:
Merge remote-tracking branch 'js_packages_tsconfig/REWRITE'

askvortsov1 committed:
Merge remote-tracking branch 'php_packages_testing/REWRITE'

askvortsov1 committed:
Merge remote-tracking branch 'php_packages_phpstan/REWRITE'

askvortsov1 committed:
Merge remote-tracking branch 'extensions_akismet/REWRITE'

askvortsov1 committed:
Merge remote-tracking branch 'extensions_approval/REWRITE'

askvortsov1 committed:
Merge remote-tracking branch 'extensions_bbcode/REWRITE'

askvortsov1 committed:
Merge remote-tracking branch 'extensions_embed/REWRITE'

askvortsov1 committed:
Merge remote-tracking branch 'extensions_emoji/REWRITE'

askvortsov1 committed:
Merge remote-tracking branch 'extensions_flags/REWRITE'

askvortsov1 committed:
Merge remote-tracking branch 'extensions_lang_english/REWRITE'

askvortsov1 committed:
Merge remote-tracking branch 'extensions_likes/REWRITE'

askvortsov1 committed:
Merge remote-tracking branch 'extensions_lock/REWRITE'

askvortsov1 committed:
Merge remote-tracking branch 'extensions_markdown/REWRITE'

askvortsov1 committed:
Merge remote-tracking branch 'extensions_mentions/REWRITE'

askvortsov1 committed:
Merge remote-tracking branch 'extensions_nicknames/REWRITE'

askvortsov1 committed:
Merge remote-tracking branch 'extensions_package_manager/REWRITE'

askvortsov1 committed:
Merge remote-tracking branch 'extensions_pusher/REWRITE'

askvortsov1 committed:
Merge remote-tracking branch 'extensions_statistics/REWRITE'

askvortsov1 committed:
Merge remote-tracking branch 'extensions_sticky/REWRITE'

askvortsov1 committed:
Merge remote-tracking branch 'extensions_subscriptions/REWRITE'

askvortsov1 committed:
Merge remote-tracking branch 'extensions_suspend/REWRITE'

askvortsov1 committed:
Merge remote-tracking branch 'extensions_tags/REWRITE'

askvortsov1 committed:
Merge remote-tracking branch 'framework_core/REWRITE'

askvortsov1 committed:
chore: remove centralizable repo config

- Remove .github conf folder from Flarum packages

- Remove styleci config from all packages

askvortsov1 committed:
chore: add monorepo config file

askvortsov1 committed:
chore: set up composer path repos

askvortsov1 committed:
chore: flarum-cli audit infra --fix

askvortsov1 committed:
chore: restore issue templates

askvortsov1 committed:
chore: move core's README to framework root.

dependabot[bot] committed:
chore(deps): bump glob-parent from 3.1.0 to 5.1.2 in /extensions/emoji/js (#3345)

* chore(deps): bump glob-parent in /extensions/emoji/js

Bumps [glob-parent](https://github.com/gulpjs/glob-parent) from 3.1.0 to 5.1.2.
- [Release notes](https://github.com/gulpjs/glob-parent/releases)
- [Changelog](https://github.com/gulpjs/glob-parent/blob/main/CHANGELOG.md)
- [Commits](https://github.com/gulpjs/glob-parent/compare/v3.1.0...v5.1.2)

---
updated-dependencies:
- dependency-name: glob-parent
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>

* Apply fixes from StyleCI

[ci skip] [skip ci]

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: StyleCI Bot <bot@styleci.io>

StyleCIBot committed:
Apply fixes from StyleCI

[ci skip] [skip ci]

askvortsov1 committed:
chore: add version to core/ext package.json

This is necessary for yarn workspaces to work.

askvortsov1 committed:
chore: use yarn v1 in extensions/core

askvortsov1 committed:
feat: use yarn workspaces

flarum-bot committed:
Bundled output for commit 5e82ded8c38f15e69c01a639d28917fe9bb29f01

Includes transpiled JS/TS.

[skip ci]

luceos committed:
fix(tags)!: reduce tag filter query from 3+ second

This PR removes the subqueries inserted for each selected tag. As this
caused the query to take over 3 seconds on communities with hundred
million posts.

```sql
and (`discussions`.`id` in (select `discussion_id` from `discussion_tag` where `tag_id` = 105))
```

The above part takes on average 3.85 seconds, with a join it takes 0.00
seconds.

BREAKING CHANGE: extending the query might have different outcome as we
now use a join and no subquery.

StyleCIBot committed:
Apply fixes from StyleCI

[ci skip] [skip ci]

luceos committed:
add distinct and clean up code further

StyleCIBot committed:
Apply fixes from StyleCI

[ci skip] [skip ci]

luceos commented:
I have tested this extensively locally and am now working towards testing this out in a staging environment.

askvortsov1 committed:
fix: Assorted Typing Fixes (#3348)

With all the commits below, we resolve all outstanding typing issues in the repo, and CI jobs run green.

* fix: Convert DashboardPage and DashboardWidget to TypeScript

* fix: fix type errors in package manager ext

* fix: Convert Post component to TypeScript

* fix: avatar typings should accept null user

* fix: convert Notification component to TypeScript

* fix: properly use `typeof` in ForumApplication

* feat: make Notification content attr generic

* chore: format Notification component

* fix: Convert DiscussionRenamedNotification to TypeScript

* fix(pusher) move shims to a location where they get applied

* fix(pusher): fix some typing errors

* fix(akismet): fix some typing issues

* chore: update core dist typings

* chore(pusher): format

* fix: anchorScroll should accept string selectors

* fix: more accurately represent ApiQueryParamsPlural

* fix: convert PostStreamState to TypeScript

* chore(core): rebuild typings

* feat: allow extending app.routes

* fix: more flexible typings for highlight.ts

* fix: use primitive `number` type for Discussion typings

* fix: convert DiscussionListItem to TypeScript

* chore: rebuild core typings

* fix: final pusher type fixes

* feat: start tags TypeScript conversion

* fix: require-dev tags in pusher for CI TypeScript purposes.

* chore(core): format

* chore(tags): build dist typings

* feat(pusher): use dist types from tags.

* feat: convert flags to TypeScript

* chore(flags): generate dist typings

* fix(akismet): last type errors

* chore: update .yarn-integrity

* chore: partially run flarum-cli audit infra --fix

The tsconfig changes from that command are ignored, since we don't yet support "replacable sections" that would let us add custom config.

* chore: use type imports

* fix: broader gitattributes

* chore: run flarum-cli audit infra --monorepo --fix

* feat: make `app.data` typings extensible

* chore(core): format

* chore: boost tags TypeScript coverage

* fix(tags): further increase type coverage.

flarum-bot committed:
Bundled output for commit a595665bfbf7820c58a3911df154ced435b6fbfd

Includes transpiled JS/TS.

[skip ci]

askvortsov1 reviewed (approved):
From a code perspective, looks good. IMO it's good to merge after your testing.

SychO9 reviewed (dismissed):
Code looks good, haven't tested locally, we should introduce integration tests for gambits.

Another improvement could be to grab all tag IDs at once instead of calling `getIdForSlug` every time (query for every tag filter), impact would only be noticeable with a large number of tag filters applied.

SychO9 committed:
fix: `Post--by-start-user` class not working (#3356)

flarum-bot committed:
Bundled output for commit 46f8cf46284754f98588646fc995fef420852a0a

Includes transpiled JS/TS.

[skip ci]

luceos commented:
> Another improvement could be to grab all tag IDs at once instead of calling `getIdForSlug` every time (query for every tag filter), impact would only be noticeable with a large number of tag filters applied.

The performance hit isn't from the tags per se, but from the subquery that selects all discussions belonging to those tags. So your optimalisation by itself would have a far less positive outcome than using a join in my improvement.

---

Hold off from merging this, I want to push this and the post number changes into one branch for testing before we merge it.

BartVB commented:
If I understand correctly Sami means moving `getIdForSlug()` out of the `foreach` (or even getting rid of the foreach altogether if possible) _and_ using a join.

The performance impact of that additional change will be much smaller, but doing a `foreach` in a query is suboptimal (to use an euphemism). Not the biggest issue in this case, but removing extraneous queries one at a time would really help in the long run. Would also set an example for extension developers.

luceos committed:
possibly fixes the issue with post number calculation

StyleCIBot committed:
Apply fixes from StyleCI

[ci skip] [skip ci]

luceos committed:
fix prefix failing in tests

StyleCIBot committed:
Apply fixes from StyleCI

[ci skip] [skip ci]

luceos committed:
missed another prefix for tests

askvortsov1 committed:
fix: canonicalize comparisons in Posts ListTest

That doesn't matter since there's no default sort for posts in the backend.

davwheat committed:
a11y: use `feed` aria role for DiscussionList (#3359)

* a11y: use `feed` aria role for DiscussionList

* a11y: add aria set info roles, move feed roles to more appropriate element

flarum-bot committed:
Bundled output for commit b95b3646d3016b5163de51c465cc33a5c2eb14c1

Includes transpiled JS/TS.

[skip ci]

askvortsov1 committed:
feat: deprecate discussion `post_number_index` field.

flarum-bot committed:
Bundled output for commit 00b91518646eeddb09635bfa76b4e92591c2595f

Includes transpiled JS/TS.

[skip ci]

luceos commented:
What would probably be a better solution would be to use an in-memory (array or replacable) cache on `getIdForSlug` altogether. That would reduce the number of queries to 1 for all similar requests over the whole product, automatically including extensions.

What do you think? I am also not sure it should be part of this PR, but it could.

SychO9 commented:
It was just a side-comment about something I noticed, let's not do anything about it in this PR 👍🏼

tankerkiller125 commented:
@luceos I've been taking a quick look at places where database caching of some kind would be beneficial, this is for sure one of the use cases where a replaceable cache would be good. (Redis or other KV store would be preferable for HA setups/reasons)

askvortsov1 commented:
Either way, I think this is good to merge (squash please!), assuming the tests you're running are going well.

luceos commented:
No confirmation from actual staging environment yet, holding off from merging.

BartVB commented:
I don't think using a cache for `getIdForSlug` would be the best way forward. Those queries are as light as they get, using Redis wouldn't help in that case because time is spent on network latency, not processing. Network latency is approx the same for a DB server and Redis.

IMO that energy could be better spent in reducing the amount of N+1 queries in Flarum. There are quite a few places where queries are run in a `foreach()`. Also in this case most of the performance increase would come from reducing the number of hits on a database/caching layer.

> let's not do anything about it in this PR

That sounds like a smart move.

The join/sorting issue
------

When it comes to the query changes in this PR; it doesn't bring the performance increase that we hoped for.  For some reason MySQL still uses `discussion_tag` as the primary table for joins which results in horrible performance when sorting.

The optimized query (which returns in a couple of milliseconds) has:
```sql
left join discussion_tag dt on (discussions.id = dt.discussion_id and tag_id = 193)
```

I had expected the optimizer to figure out that both are the same, but apparently it doesn't (at least not on Percona 8.0.26).

The queries produced by this PR contain:
```sql
left join `discussion_tag` on `discussions`.`id` = `discussion_tag`.`discussion_id`
[...]
and (`discussion_tag`.`tag_id` = 105)
```

This is the first row of an EXPLAIN:

```
           id: 1                                                                                                                  
  select_type: PRIMARY
        table: discussion_tag
   partitions: NULL  
         type: ref                                               
possible_keys: PRIMARY,discussion_tag_tag_id_foreign
          key: discussion_tag_tag_id_foreign
      key_len: 4                                                                                                                  
          ref: const
         rows: 90522 
     filtered: 100.00                                                                                                             
        Extra: Using where; Using index; Using temporary; Using filesort
```

If it's using `discussion_tag` then it can't use `discussions_is_sticky_last_posted_at_index` for sorting. That means that it needs to create a temporary table with all results (90k rows in this case) and then sort them manually.

I can think of some SQL hacks to work around this, but I'm hoping that there is an Eloquent way to do this?

SychO9 commented on extensions/tags/src/Query/TagFilterGambit.php:
```diff
@@ -55,23 +55,23 @@ protected function constrain(Builder $query, $rawSlugs, $negate)
+                            $negate ? '!=' : '=',
+                            $id
+                        );
+                    }
```
Let's try this instead:
```suggestion
        $query
            ->distinct()
            ->join('discussion_tag', function (\Illuminate\Database\Query\JoinClause $join) use ($slugs, $negate) {
                $join
                    ->on('discussions.id', '=', 'discussion_tag.discussion_id')
                    ->where(function (\Illuminate\Database\Query\JoinClause $join) use($slugs, $negate) {
                        foreach ($slugs as $slug) {
                            if ($slug === 'untagged' && ! $negate) {
                                $join->orWhereNull('discussion_tag.tag_id');
                            } elseif ($slug === 'untagged' && $negate) {
                                $join->orWhereNotNull('discussion_tag.tag_id');
                            } elseif ($id = $this->tags->getIdForSlug($slug)) {
                                $join->orWhere(
                                    'discussion_tag.tag_id',
                                    $negate ? '!=' : '=',
                                    $id
                                );
                            }
                        }
                    });
            });
```

BartVB replied:
Alas that doesn't result in a significant speedup. But if I change `->join()` into `->leftJoin()` then the request times drop dramatically. Seems to be doing what we want!

As said; I'm not exactly proficient with Eloquent but can't the `->where()` be replaced with `->on()`? That way `\Illuminate\Database\Query\JoinClause` isn't needed. That would be a bit more readable/cleaner because in the end it's not really a `WHERE` but a part of the `ON` part of that `LEFT JOIN`.

Great work!

SychO9 reviewed (commented):
yeah, that's pretty bad, it traverses all 90k rows (all existing rows) for the results.

I tested and suggested a block of code that should apply the desired query. However, there is actually an issue I noticed about using a join, if you search with multiple tag gambits/filters, it will attempt to apply the same join again, which means join column conditions won't be unique with the `discussion_tag` table having been previously joined. We need to use a unique table alias every time.

davwheat committed:
feat: allow multiple children to be passed to ConfirmDocumentUnload component (#3372)

flarum-bot committed:
Bundled output for commit 803f4cc792a1cd73e6fdb829d8f20e9ccd032722

Includes transpiled JS/TS.

[skip ci]

dsevillamartin committed:
fix(subscriptions): filter followed discussions when searching (#3376)

Fixes #3375

flarum-bot committed:
Bundled output for commit cd8ed6cf66789b187873072dabc43ce873493243

Includes transpiled JS/TS.

[skip ci]

SychO9 committed:
feat: Graceful frontend extension initialization failure (#3349)

flarum-bot committed:
Bundled output for commit ca7055f5d04af64755aef159d288e9337f17e885

Includes transpiled JS/TS, and Typescript declaration files (typings).

[skip ci]

davwheat committed:
chore: don't track `.yarn-integrity` in git (#3374)

flarum-bot committed:
Bundled output for commit 7c3da27fb32db6941f6401fee8b1727d3efd3ef4

Includes transpiled JS/TS, and Typescript declaration files (typings).

[skip ci]

luceos committed:
Update extensions/tags/src/Query/TagFilterGambit.php

Co-authored-by: Sami Mazouz <sychocouldy@gmail.com>

luceos committed:
clean up code and use leftJoin

StyleCIBot committed:
Apply fixes from StyleCI

[ci skip] [skip ci]

askvortsov1 reviewed (approved):
LGTM! Please squash when merging

SychO9 reviewed (changes requested):
We can't merge this yet until we figure out how best to uniquely alias the joined table (try applying the filter/gambit multiple times to see what I mean `tag:support tag:feedback`)

> However, there is actually an issue I noticed about using a join, if you search with multiple tag gambits/filters, it will attempt to apply the same join again, which means join column conditions won't be unique with the discussion_tag table having been previously joined. We need to use a unique table alias every time.

luceos commented:
> We can't merge this yet until we figure out how best to uniquely alias the joined table (try applying the filter/gambit multiple times to see what I mean `tag:support tag:feedback`)
> 
> > However, there is actually an issue I noticed about using a join, if you search with multiple tag gambits/filters, it will attempt to apply the same join again, which means join column conditions won't be unique with the discussion_tag table having been previously joined. We need to use a unique table alias every time.

Yeah we noticed it's broken. Will fix asap. Sorry about that.

imorland committed:
feat: add discussion sortmap to container binding (#3377)

flarum-bot committed:
Bundled output for commit 1d27f62c15b569f4cd74a24fc38a636d34933f9f

Includes transpiled JS/TS, and Typescript declaration files (typings).

[skip ci]

davwheat committed:
fix: notification timestamp mismatch in specific configurations (#3379)

When app's PHP server's timezone is not set to UTC, which would normally be done by Flarum's bootstrapping process, this call to Carbon would return a different timestamp to other calls elsewhere in Flarum, causing issues with notifications seemingly being delivered too early/late rather than at the actual time they were triggered.

flarum-bot committed:
Bundled output for commit 9f7c5defaa1d082bda9af99dfc84d08e3e512ca5

Includes transpiled JS/TS.

[skip ci]

askvortsov1 committed:
fix: dont ignore optional dependencies on disabled extensions. (#3352)

There is a check in the ExtensionManager::resolveExtensionOrder function that ignores optional dependencies on extensions that don't exist in the system. This is sufficient for resolution purposes.

The filter removed in this PR would ignore optional dependencies on non-enabled extensions, so when such an extension was enabled, dependency resolution would run incorrectly.

flarum-bot committed:
Bundled output for commit 431ba304348f0f923c65183e5f8e3c03ac3da9b9

Includes transpiled JS/TS, and Typescript declaration files (typings).

[skip ci]

SychO9 committed:
fix: extension listeners added before core listeners (#3373)

flarum-bot committed:
Bundled output for commit 98e607a7041b55130cfc5d0c8fb3e23b4b64e951

Includes transpiled JS/TS.

[skip ci]

imorland committed:
fix: extensions not being able to use the renderer without context (#3382)

flarum-bot committed:
Bundled output for commit b0fa795cc936a5538af892312e97bbf763625e18

Includes transpiled JS/TS.

[skip ci]

SychO9 committed:
fix: languages with combining characters cannot be searched (#3385)

* test: searching other types of languages
* fix: languages with combining characters cannot be searched
* test: can search in discussion titles

flarum-bot committed:
Bundled output for commit 6de1ea0194ac6c40e5775ffe9d8b7890c458c9b1

Includes transpiled JS/TS, and Typescript declaration files (typings).

[skip ci]

imorland committed:
feat: export `utils/isSafariMobile` (#3389)

flarum-bot committed:
Bundled output for commit 542cc42538d8f75f58442f0d4b45a0dcbda5ad4b

Includes transpiled JS/TS.

[skip ci]

dsevillamartin committed:
feat(emoji): support emoji 14 & unicode 14 (#3378)

flarum-bot committed:
Bundled output for commit c658eb10f4ac0c22e068827af462282a9a0c9e8a

Includes transpiled JS/TS.

[skip ci]

SychO9 committed:
fix: slug transliteration assuming the language is always `en` (#3387)

* fix: transliteration assuming the language is always `en`
* test: transliteration works with forum locale

flarum-bot committed:
Bundled output for commit fb717db57cc53d44d064d34b8aacb71ebca47322

Includes transpiled JS/TS, and Typescript declaration files (typings).

[skip ci]

SychO9 committed:
fix: PHP 8.1 deprecation warnings (#3384)

flarum-bot committed:
Bundled output for commit be2da3e12fed5f3e6a1168c29314c35ac4b09141

Includes transpiled JS/TS, and Typescript declaration files (typings).

[skip ci]

askvortsov1 committed:
fix(tags): TagDiscussionModal don't go to null tag (#3404)

If the tag we are trying to select/highlight is undefined, we should not run the scroll animation.

flarum-bot committed:
Bundled output for commit 4cb15501a253aedc841f68852319edbf3943b48e

Includes transpiled JS/TS, and Typescript declaration files (typings).

[skip ci]

askvortsov1 committed:
fix: don't duplicate path in URL to post on subdirectory installs. (#3354)

`baseUrl` includes the path portion of the url, as does `basePath`, which is included in `app.route.x` function outputs. The `baseOrigin` (with no path component at all) should be used instead.

flarum-bot committed:
Bundled output for commit 863a8ddfffd7e56ed8cb34220d8342b1cff53c43

Includes transpiled JS/TS, and Typescript declaration files (typings).

[skip ci]

askvortsov1 committed:
fix(core): don't pass null to functions where that's deprecated (#3405)

Should fix remaining PHP 8.1 deprecation warnings/

SychO9 committed:
chore: reflect our issue tracking protocol in issue templates. (#3401)

SychO9 committed:
chore: fix package manager tests (#3407)

luceos committed:
potential fix for table alias

luceos committed:
Merge branch 'main' into dk/20220318-tag-index

StyleCIBot committed:
Apply fixes from StyleCI

[ci skip] [skip ci]

luceos commented:
@SychO9 I've aliased the table, should be unique for whatever passthrough. Unless the same slugs were used twice.

luceos commented:
According to @BartVB we cannot merge this yet.

BartVB commented:
This issue can be reduced to:

```sql
select discussions.id 
  from discussions 
    left join discussion_tag on discussions.id = discussion_tag.discussion_id
  where discussion_tag.tag_id = 105 
  order by is_sticky desc, last_posted_at desc
  limit 21;
```
Which is slow.

Where it went wrong is that `LEFT JOIN on tag_id = 105` is not the same as `LEFT JOIN ... WHERE tag_id = 105`. In the first instance you would of course get all rows from the discussions table, rows where `tag_id=105` get extra information from the `discussion_tag` table added. That's quick because MySQL hardly needs to filter so it uses `discussions` as the first join table. Because of that it can use an index for sorting. Problem is that it will not return the correct results.

If you add the `where` clause or use an `inner join` then you'll get the correct results, but `tag_id` has low cardinality so MySQL first takes matching rows from `discussion_tag` and then joins that with `discussions`. That would make sense if we didn't need to sort... Without the index quite a few rows need to be sorted, depending on how popular a tag is. I.e. if we change the query to `WHERE tag_id = 206` then the query returns instantly because that tag_id only has 6 discussions, instead of the 123.101 discussions in tag_id = 105.

So in a lot of cases (where there are no a lot of rows in a tag) the MySQL optimiser choses right.

It's possible to force MySQL to use the `discussions_is_sticky_last_posted_at_index` index. For instance with a STRAIGHT_JOIN:

```sql
select discussions.id 
  from discussions 
    straight_join discussion_tag on discussions.id = discussion_tag.discussion_id
  where discussion_tag.tag_id = 105 
  order by is_sticky desc, last_posted_at desc
  limit 21;
```

or

```sql
select discussions.id 
  from discussions force index (discussions_is_sticky_last_posted_at_index)
    left join discussion_tag on discussions.id = discussion_tag.discussion_id
  where discussion_tag.tag_id = 105 
  order by is_sticky desc, last_posted_at desc
  limit 21;
```

But doing that would make queries slow for queries that involve tags with only a few discussions because a very large part of the index needs to be read and skipped in those cases.

So it's back to the drawing board it seems...
